### PR TITLE
Add explicit message when `noexec` prevents library loading.

### PR DESCRIPTION
**Pull request by mkw**
Add explicit message when `noexec` prevents library loading.

Motivation:

Docker's `--tmpfs` flag mounts the temp volume with `noexec` by default, resulting in an UnsatisfiedLinkError.  While this is good security practice, it is a surprising failure from a seemingly innocuous flag.

Modifications:

Add a best-effort attempt to detect the `noexec` flag in `NativeLibraryLoader`.  This solution is (probably) limited to Oracle/OpenJDK VM's and requires Java 7 or higher, and also requires `--add-opens=java.base/sun.nio.fs=...` for Java 9.  It avoids error message parsing and it does not directly attempt to read mount information.

Also upgraded animal-sniffer-maven-plugin and added configuration to make it ignore code annotated with `io.netty.util.SuppressForbidden` to address @Scottmitch's feedback from original PR.

Result:

Fixes [#6678].

### CODE REVIEW

**Comment by mkw** [comment]
I accidentally closed #6685 cleaning up branches in our clone.  This is a second attempt and addresses the comments that @Scottmitch had on that PR, being:

- Instead of using reflection for Java7 API's, it checks the java version in a guard clause and directly uses Files/FileStore/Path.  This requires either ignoring _all_ references to these fairly generic NIO APIs in Java7 or upgrading sniffer and using annotations for excluding the method.  I chose the latter since my first shot at the former did not work, and it was not clear why.
- It documents the InaccessibleObjectException expected when Jigsaw blocks access to the private API's used to extract mount information.
- It removes a redundant throw/catch when checking for noexec.

File: common/src/main/java/io/netty/util/internal/NativeLibraryLoader.java [hunk 1]
```diff
@@ -368,4 +385,59 @@ private static void closeQuietly(Closeable c) {
...
+                } catch (Throwable t) {
+                    logger.debug("Failed to initialize NativeLibraryLoader.NoexecFileStoreDetector", t);
+                }
+            }
+            MOUNT_ENTRY_METHOD = mountEntryMethod;
+            HAS_OPTION_METHOD = hasOptionMethod;
+        }
+
+        @SuppressForbidden(reason = "to obtain mount information about file stores")
+        private static boolean isOnNoexecFileStore(File file)
```
Comment (normanmaurer) [inline common/src/main/java/io/netty/util/internal/NativeLibraryLoader.java]: remove "private" to make it package-private

File: common/src/main/java/io/netty/util/internal/NativeLibraryLoader.java [hunk 2]
```diff
@@ -368,4 +385,59 @@ private static void closeQuietly(Closeable c) {
...
+                throws IOException, InvocationTargetException, IllegalAccessException {
+            if (PlatformDependent.javaVersion() < 7 || HAS_OPTION_METHOD == null || MOUNT_ENTRY_METHOD == null) {
```
Comment (normanmaurer) [inline common/src/main/java/io/netty/util/internal/NativeLibraryLoader.java]: you can remove `PlatformDependent.javaVersion() < 7` as these methods will be null in this case anyway.

File: pom.xml [hunk 3]
```diff
@@ -706,7 +706,7 @@
              be used even when compiling with java 1.7+ -->
         <groupId>org.codehaus.mojo</groupId>
         <artifactId>animal-sniffer-maven-plugin</artifactId>
-        <version>1.11</version>
+        <version>1.15</version>
```
Comment (normanmaurer) [inline pom.xml]: why this change ?

File: pom.xml [hunk 4]
```diff
@@ -721,6 +721,9 @@
             <artifactId>java16</artifactId>
             <version>1.1</version>
           </signature>
+          <annotations>
+            <annotation>io.netty.util.SuppressForbidden</annotation>
```
Comment (normanmaurer) [inline pom.xml]: Why this change ?

File: common/src/main/java/io/netty/util/internal/NativeLibraryLoader.java [hunk 5]
```diff
@@ -368,4 +385,59 @@ private static void closeQuietly(Closeable c) {
...
+                return false;
+            }
+            FileStore fileStore = Files.getFileStore(file.toPath());
+            Object unixMountEntry = MOUNT_ENTRY_METHOD.invoke(fileStore);
+            return (Boolean) HAS_OPTION_METHOD.invoke(unixMountEntry, "noexec");
```
Comment (Scottmitch) [inline common/src/main/java/io/netty/util/internal/NativeLibraryLoader.java]: I guess there is no way to detect an issue by avoiding internal APIs?
- Checking if the directory is readable and/or [File#canExecute](https://docs.oracle.com/javase/7/docs/api/java/io/File.html#canExecute())
- Use [FileStore#getAttribute](https://docs.oracle.com/javase/7/docs/api/java/nio/file/FileStore.html#getAttribute(java.lang.String)) or some other attribute API offered by NIO

File: common/src/main/java/io/netty/util/internal/NativeLibraryLoader.java [hunk 6]
```diff
@@ -15,18 +15,22 @@
...
+import java.io.Closeable;
 import java.io.File;
 import java.io.FileOutputStream;
 import java.io.IOException;
 import java.io.InputStream;
 import java.io.OutputStream;
+import java.lang.reflect.InvocationTargetException;
 import java.lang.reflect.Method;
 import java.net.URL;
+import java.nio.file.FileStore;
```
Comment (normanmaurer) [inline common/src/main/java/io/netty/util/internal/NativeLibraryLoader.java]: @johnou good point!

@mkw we need to use reflection here or move it to another class that is only loaded if java version >= 7

File: pom.xml (hunk 3 above)
Comment (mkw) [inline pom.xml]: My mistake -- undoing -- thought it was needed for annotation-based suppression (but I'm also about to undo that).

File: pom.xml (hunk 4 above)
Comment (mkw) [inline pom.xml]: Idea is to more finely target suppression of the animal sniffer errors, but I agree that reflection is better; undoing and fixing.

File: common/src/main/java/io/netty/util/internal/NativeLibraryLoader.java (hunk 5 above)
Comment (mkw) [inline common/src/main/java/io/netty/util/internal/NativeLibraryLoader.java]: I didn't realize it, but it appears that [`File#canExecute`](https://docs.oracle.com/javase/7/docs/api/java/io/File.html#canExecute()) _can_ be used for this: it does require that I temporarily try to update the POSIX permissions, but that is far preferable to the internal API mess that I'd created.  Expect an updated PR soon.

**Comment by mkw** [comment]
Sorry for the long delay on my end, as well.  The revised version uses only public Java API's and reflection to avoid problems with JDK6.

Previously, @Scottmitch had asked to avoid reflection, but so far as I can tell I can only suppress the warnings from Animal Sniffer by suppressing warnings about _all_ uses of classes like `java.nio.file.Files`, which seems overly broad.  Note that the Animal Sniffer Maven plugin does support a custom "suppress only here" annotation, but it is not configured in Netty.  Were it to be configured, I suspect that the class list in the main pom.xml should be mostly dropped and replaced with annotations across the project.

File: common/src/main/java/io/netty/util/internal/NativeLibraryLoader.java [hunk 7]
```diff
@@ -371,4 +390,62 @@ private static void closeQuietly(Closeable c) {
...
         // Utility
     }
+
+    private static final class NoexecVolumeDetector {
+
+        private static boolean canExecuteExecutable(File file)
+                throws ClassNotFoundException, NoSuchMethodException, InvocationTargetException,
+                       IllegalAccessException {
+            if (PlatformDependent.javaVersion() < 7) {
+                // Assume true because we
```
Comment (Scottmitch) [inline common/src/main/java/io/netty/util/internal/NativeLibraryLoader.java]: incomplete comment ... please complete this comment or remove it.

File: common/src/main/java/io/netty/util/internal/NativeLibraryLoader.java [hunk 8]
```diff
@@ -371,4 +390,62 @@ private static void closeQuietly(Closeable c) {
...
+            // If we can already execute, there is nothing to do.
+            if (file.canExecute()) {
+                return true;
+            }
+
+            // On volumes, with noexec set, even files with the executable POSIX permissions will fail to execute.
+            // So, check for executable permissions, then find out if the file can be executed.
+
+            // Everything is done reflectively to prevent the animal-sniffer plugin from complainin.
+            Class<?> filesClass = Class.forName("java.nio.file.Files");
```
Comment (Scottmitch) [inline common/src/main/java/io/netty/util/internal/NativeLibraryLoader.java]: instead of using reflection to load these resources can we reference the class directly? We just need to be careful about class loading on older JDKs.

```java
private static boolean canExecuteFile(File f) {
  return PlatformDependent.javaVersion() < 7 ? true : Java7FilesUtil. canExecuteFile(f);
}

// In another class file ... Java7FilesUtil.java
final class Java7FilesUtil {
  static boolean canExecuteFile(File f) {
    ...
  }
}
```

Comment (mkw) [inline common/src/main/java/io/netty/util/internal/NativeLibraryLoader.java]: First, if I'm reading, `io.netty.util.internal.PlatformDependent#newConcurrentDeque` correctly, it _appears_ that it is fine to use Java7 classes provided that they are protected by a version check -- no?

Whether or not this is true, I would be happy to do this non-reflectively, but doing so would require adding the classes to the Animal Sniffer configuration.  My understanding of how Animal Sniffer works is that doing so would suppress the usages of these classes globally within the Netty codebase.  Given how natural it has been for developers to use `java.nio.file.*`, this seems unwise.

As of [Animal Sniffer 1.11](http://blog.soebes.de/blog/2014/03/25/animal-sniffer-version-1-dot-11-released/), it is possible to configure an annotation to suppress warnings in specific classes or method, but so far as I can tell, this is not yet configured for the project.  If you want, I could make a separate issue that extended `io.netty.util.SuppressForbidden` or added something similar (`io.netty.util.IgnoreJDK7` and `io.netty.util.IgnoreJDK8`?), added these to relevant scopes, and removed the global class exclusions from the sniffer configuration.

Please advise.

File: common/src/main/java/io/netty/util/internal/NativeLibraryLoader.java (hunk 7 above)
Comment (mkw) [inline common/src/main/java/io/netty/util/internal/NativeLibraryLoader.java]: Sorry -- was rushing yesterday and didn't re-read my comments.  They should be much clearer now.

File: common/src/main/java/io/netty/util/internal/NativeLibraryLoader.java (hunk 8 above)
Comment (Scottmitch) [inline common/src/main/java/io/netty/util/internal/NativeLibraryLoader.java]: > it appears that it is fine to use Java7 classes provided that they are protected by a version check -- no?

No. If you need to use features not available in JDK6 you need a version check. Netty's minimum compatibility is currently at JDK6 so it must run against this JDK and we can't attempt to load any classes which won't exist in this context.

> My understanding of how Animal Sniffer works is that doing so would suppress the usages of these classes globally within the Netty codebase ... If you want, I could make a separate issue

I would prefer to have the code reference actual types and update the Animal sniffer configuration to provide more fine grained exceptions. For the purposes of this PR it is fine to add an exception using the current approach and we can update the animal sniffer exceptions in a followup PR.

Comment (normanmaurer) [inline common/src/main/java/io/netty/util/internal/NativeLibraryLoader.java]: The important here is that you "MUST" import the java7 classes in another class that you only access if the version check match. Otherwise you will get a classloader exception when try to use java6

Comment (mkw) [inline common/src/main/java/io/netty/util/internal/NativeLibraryLoader.java]: I'm going to apply this rule, but this block, inside`io.netty.util.internal.PlatformDependent` confuses me:
```
    /**
     * Returns a new concurrent {@link Deque}.
     */
    public static <C> Deque<C> newConcurrentDeque() {
        if (javaVersion() < 7) {
            return new LinkedBlockingDeque<C>();
        } else {
            return new ConcurrentLinkedDeque<C>();
        }
    }
```
The Javadoc of `java.util.concurrent.ConcurrentLinkedDeque` includes `@since 1.7`, and it is imported on line 42.

Assuming the rule is "you 'MUST' import the java7 classes in another class that you only access if the version check match," shouldn't PlatformDependant fail to load entirely on JDK6?  Or am I misunderstanding something?

Comment (Scottmitch) [inline common/src/main/java/io/netty/util/internal/NativeLibraryLoader.java]: You found a bug (/cc @normanmaurer) 👍 

This should be:

```java
return new java.util.concurrent.ConcurrentLinkedDeque<C>();
```

Comment (normanmaurer) [inline common/src/main/java/io/netty/util/internal/NativeLibraryLoader.java]: will look into it

Comment (normanmaurer) [inline common/src/main/java/io/netty/util/internal/NativeLibraryLoader.java]: Fixed in https://github.com/netty/netty/pull/6822

Comment (mkw) [inline common/src/main/java/io/netty/util/internal/NativeLibraryLoader.java]: @Scottmitch The non-reflective version is in, though it requires suppressing checks for uses of all of:

* java.nio.file.attribute.PosixFilePermission
* java.nio.file.Files
* java.nio.file.LinkOption
* java.nio.file.Path
* java.io.File (so it can call `File#toPath()`)

Note that the research in [this repository](https://github.com/mkw/java6test/tree/master) seems to have convinced @normanmaurer that the Java7 check at the beginning of the method should work (see the comments on #6822) in Java6 with the imports present.

**Comment by Scottmitch** [comment]
4.1 (ed5fcbb77) 4.0 (aff9e7ac0)

**Comment by Scottmitch** [comment]
thanks @mkw !
